WordPress/openverse
# Create filtered index before promoting primary index during data refresh

## Fixes



Fixes #2981 by @AetherUnbound 

## Description



This PR updates the data refresh DAG to run the steps to create the filtered index _before_ promoting the primary index . The goal is to prevent API instability during data refreshes, potentially caused by the API using an ES index that is also being used as the source of a reindex (to create the filtered index).

To accomplish this, I refactored the `create_filtered_index` DAG factory to add a new factory that generates a `create_filtered_index` TaskGroup and a separate `promote_filtered_index` TaskGroup (that does the index readiness check, promotion, and deletion). The data refresh DAGs now, rather than triggering the separate `create_filtered_index` DAGs, use this factory method to create the taskgroups and add them directly into its own flow:


<img width="1135" alt="Screenshot 2023-11-01 at 2 23 31 PM" src="https://github.com/WordPress/openverse/assets/63313398/12208425-2124-450a-bf3e-267c6001868b">

_Collapsed view, showing that we complete `ingest_upstream`, wait for the index readiness check on the primary index, then **create the filtered index before promoting anything**, and only then do we promote the primary index followed by the filtered index._


<img width="987" alt="Screenshot 2023-11-01 at 2 23 52 PM" src="https://github.com/WordPress/openverse/assets/63313398/c0a6178a-7aa4-434c-b16c-2bc5b90b4ac0">

_Expanded view of the filtered index creation_

<img width="1022" alt="Screenshot 2023-11-01 at 2 24 06 PM" src="https://github.com/WordPress/openverse/assets/63313398/a900fbd9-18f0-457b-b811-bf1a23a8bf5b">

_Expanded view of the filtered index promotion_

We keep the separate filtered index creation DAGs as these are still useful for running manually when testing filtered indices or updating the sensitive terms list. These DAGs now just reuse the same task groups and add the concurrency check. Because the DAG is _only_ run manually, I removed the `force` parameter to break through the concurrency check.

<img width="1091" alt="Screenshot 2023-11-01 at 2 24 38 PM" src="https://github.com/WordPress/openverse/assets/63313398/7ddfcdeb-2502-46c4-a9fe-e0be7a013f13">

_New view of the create filtered index DAG_

## Testing Instructions


Test the data refreshes locally. Also test the filtered index DAGs: I tried running the DAG locally with `origin_index_suffix` empty (so it would use the current index) and `destination_index_suffix` set to `foo`. After the DAG completed I checked Elasticvue to see that I now had an `audio-foo-filtered` index replacing the old filtered index :)

Test that the concurrency checks work to prevent either a data refresh or filtered index dag from running at the same time as its counterpart. You can do this by triggering both at once in the shell:

```
$ just catalog/shell
> airflow dags trigger image_data_refresh && airflow dags trigger create_filtered_image_index
```

Try swapping the order of the two commands as well to test the opposite direction. (If the filtered index Dag starts second, it should fail immediately. If the data refresh starts second, it should wait for the filtered index dag. This behavior was not changed.)

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
